### ISSUE TITLE
Support dispatch by tag on the same channel group on the server side

### ISSUE THREAD

**Issue by absolute8511** [absolute8511/nsq#28]
Currently, we dispatch channel message to the client by select on the same go chan which result in the average dispatch across all the clients under the same channel.

We add a new dispatch policy which can dispatch by the tag in the message. For a client that identify its `desired_tag` = `tagA`, we try dispatch messages with `tagA` to this client only. If no any matching client, we dispatch the message to the client without any `desired_tag`.

First, we need support add extend data in the message. We allow new created topic can support tag extend feature.
The topic with extend feature can pub the message with tag as the extend data. The channel consumer can use the identify command to set the `desired_tag` as dispatch policy. While delivering the server should dispatch the messages based on the tag in the message and the `desired_tag` in the consumer.

**Comment by absolute8511** [comment]
### Extend TCP PUB protocol
```
PUB <topic_name> <topic_partition> <message_tag>\n
[ 4-byte size in bytes ][ N-byte binary data ]
 
PUB_TRACE <topic_name> <topic_partition> <message_tag>\n
[ 4-byte size in bytes ][8-byte size trace id][ N-byte binary data ]
 
<topic_name> - a valid string
<message_tag> - a valid string, only [a-zA-Z0-9] allowed, the max length is 100 bytes.
```

### Extend HTTP API
```
curl -d "<message>" http://127.0.0.1:4151/pub?topic=name&tag=xxx
```

### Message Format Extend:
```
[x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x][x]...[x][x][x][x]...
|       (int64)        ||    ||      (hex string encoded in ASCII)           || ||    ||   ext string   ||  (binary)
|       8-byte         ||    ||                 16-byte                      || ||    ||    extN-byte   ||   N-byte
---------------------------------------------------------------------------------------------------------------------------------...
  nanosecond timestamp    ^^                   message ID                      ^    ^      extend data              message body
                       (uint16)                                            uint8,  uint16
                        2-byte                                             1-byte, 2-byte
                       attempts                                     ver of extend, length of extend
```

For this tag data, use ver of extend as `0x2`,  length of extend as length of tag data, extend data as tag raw data.

### IDENTIFY on consume Extend:
Add new json field `desired_tag` while sending identify data.
`desired_tag` the tag for matching for this client.
rules:
- if `desired_tag` is not empty, but the topic do not support tag extend , return the error: tag extend on the topic is not enabled. 
- if `desired_tag` == "", all un-dispatched messages will be send to these clients.

### Tag Matching rules

- message with no tag -> delivery to the channel client with empty `desired_tag` identified.
- message with a tag -> only delivery to the channel client with the same `desired_tag` identified, and if failed to find any matching clients, delivery to the client with empty `desired_tag`.
- a message failed to match any client-> stay in the waiting list and wait client.

### Lookup create topic API
```
curl -X POST "http://127.0.0.1:4161/topic/create?topic=xxx&partition_num=2&replicator=3&suggestload=0&syncdisk=1000&extend=true"
```

### stats API and nsqadmin
- topic stats should include the flag indicated whether the extend feature is enabled
-  show the client `desired_tag` data under the channel page in the nsqadmin

**Comment by absolute8511** [comment]
solved in #34
